8294087: RISC-V: RVC: Fix a potential alignment issue and add more alignment assertions for the patchable calls/nops

zhengxiaolinX opened the pull request:
With RVC turning on, we must carefully keep all runtime-patchable instructions aligned. Code is running at full speed, when patching unaligned instructions spanning cache lines, concurrency issues may occur. This patch fixes a potential alignment issue of the patchable nop after MachUEPNode, with adding some strong alignment assertions as well. (In fact, currently the `nop` in the Verified Entry Point is fortunately aligned to 4 under RVC even without this patch, so this patch doesn't change program behaviors.)

Tested hotspot tier1 and tier2 together with other patches on QEMU.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294087](https://bugs.openjdk.org/browse/JDK-8294087): RISC-V: RVC: Fix a potential alignment issue and add more alignment assertions for the patchable calls/nops


### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [8ab3377d](https://git.openjdk.org/jdk/pull/10370/files/8ab3377d720b6c933fd85104b0a8bea98132d8ab)
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Author)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10370/head:pull/10370` \
`$ git checkout pull/10370`

Update a local copy of the PR: \
`$ git checkout pull/10370` \
`$ git pull https://git.openjdk.org/jdk pull/10370/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10370`

View PR using the GUI difftool: \
`$ git pr show -t 10370`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10370.diff">https://git.openjdk.org/jdk/pull/10370.diff</a>

</details>

RealFYang commented:
Can you add in the code alignment assertions introduced by your new RVC proposal [1]?

[1] https://github.com/zhengxiaolinX/jdk/commit/26874527d196cd6c31b1649c8c00ed4f5a6b154a

zhengxiaolinX commented:
> Can you add in the code alignment assertions introduced by your new RVC proposal [1]?
> 
> [1] [zhengxiaolinX@2687452](https://github.com/zhengxiaolinX/jdk/commit/26874527d196cd6c31b1649c8c00ed4f5a6b154a)

Of course, I have added that and modified the title and descriptions of this PR.

zhengxiaolinX commented on src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:
```diff
@@ -91,6 +91,14 @@ int MacroAssembler::align(int modulus, int extra_offset) {
   return (int)(offset() - before);
 }
 
+void MacroAssembler::assert_alignment() {
```
Thank you, that is reasonable and I have modified it; but due to the introduction of `NativeInstruction::instruction_size` in the `macroassembler_riscv.hpp`, it seems we also need to introduce `nativeinst_riscv.hpp` in it. I was wondering is this okay, or other plans?

zhengxiaolinX replied:
I have updated a revertible patch to see which code style is better.

shipilev commented on src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:
```diff
@@ -2839,6 +2847,7 @@ address MacroAssembler::trampoline_call(Address entry) {
   }
 
   address call_pc = pc();
+  assert(entry.rspec().type() == relocInfo::runtime_call_type || is_aligned(call_pc, 4), "bad alignment for patchable calls");
```
"4" is still here?

zhengxiaolinX replied:
Nice catch, modified.

shipilev reviewed (commented):
I think `static void assert_alignment(address pc, int alignment = NativeInstruction::instruction_size);` is cleaner, even though it requires a new include in `macroAssembler_riscv.hpp`.

zhengxiaolinX commented:
> I think `static void assert_alignment(address pc, int alignment = NativeInstruction::instruction_size);` is cleaner, even though it requires a new include in `macroAssembler_riscv.hpp`.

Thanks for the determination. I have reverted the revertible one.

shipilev commented on src/hotspot/cpu/riscv/nativeInst_riscv.cpp:
```diff
@@ -265,6 +265,13 @@ void NativeJump::verify() { }
+  // alignment restrictions are true. These guarantees check those
+  // conditions.
+
+  // Must be 4 byte aligned
+  MacroAssembler::assert_alignment(verified_entry);
```
So, if the intent to pass explicit `NativeInsn::insn_size` everywhere we say "Must be 4 bytes aligned", then we should probably pass it here too. Or, drop the explicit `NativeInsn::insn_size` everywhere else?

shipilev replied:
Sorry, I misread the patch, and confused `align` and `assert_alignment`.

zhengxiaolinX replied:
> Sorry, I misread the patch, and confused `align` and `assert_alignment`.

That's nothing. Thanks!

RealFYang replied:
I think this comment should be: "// Must be 4 bytes aligned"

zhengxiaolinX replied:
> I think this comment should be: "// Must be 4 bytes aligned"

Oh thanks for catching that. With hindsight, I found Aleksey had already hinted at it by indicating `...everywhere we say "Must be 4 bytes aligned"...`, but I failed to catch that :-) Thank you both!

RealFYang commented on src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:
```diff
@@ -91,6 +91,14 @@ int MacroAssembler::align(int modulus, int extra_offset) {
+void MacroAssembler::assert_alignment(address pc, int alignment) {
+  assert(is_aligned(pc, alignment), "bad alignment");
+}
+
```
Would you mind a few more tweaks?
I don't think we need two versions here. Maybe we can only keep the second one and remove the first one here?
Then the call sites will look more consistent passing an explict 'pc' argument.
Also we should move the function definition to macroAssembler_riscv.hpp.

zhengxiaolinX replied:
Of course, revised and add `static inline void` for it in the hpp file.

feilongjiang commented on src/hotspot/cpu/riscv/c1_LIRAssembler_riscv.cpp:
```diff
@@ -1372,7 +1372,7 @@ void LIR_Assembler::ic_call(LIR_OpJavaCall* op) {
 void LIR_Assembler::emit_static_call_stub() {
   address call_pc = __ pc();
-  assert((__ offset() % 4) == 0, "bad alignment");
+  __ assert_alignment(call_pc);
```
As `assert_alignment` is defined as a  static method, there is no need for `__`. `MacroAssembler::assert_alignment(call_pc)` would be better.

zhengxiaolinX replied:
Reasonable; all done.

feilongjiang commented on src/hotspot/cpu/riscv/gc/shared/barrierSetAssembler_riscv.cpp:
```diff
@@ -262,7 +262,7 @@ void BarrierSetAssembler::nmethod_entry_barrier(MacroAssembler* masm, Label* slo
     __ bind(local_guard);
 
-    assert(__ offset() % 4 == 0, "bad alignment");
+    __ assert_alignment(__ pc());
```
ditto

feilongjiang commented on src/hotspot/cpu/riscv/riscv.ad:
```diff
@@ -1318,6 +1318,7 @@ void MachPrologNode::emit(CodeBuffer &cbuf, PhaseRegAlloc *ra_) const {
 
   // insert a nop at the start of the prolog so we can patch in a
   // branch if we need to invalidate the method later
+  __ assert_alignment(__ pc());
```
ditto

feilongjiang commented on src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:
```diff
@@ -1140,6 +1140,7 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
     int vep_offset = ((intptr_t)__ pc()) - start;
 
     // First instruction must be a nop as it may need to be patched on deoptimisation
+    __ assert_alignment(__ pc());
```
ditto

feilongjiang commented on src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp:
```diff
@@ -1291,6 +1292,7 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
 
   // If we have to make this method not-entrant we'll overwrite its
   // first instruction with a jump.
+  __ assert_alignment(__ pc());
```
ditto

zhengxiaolinX commented:
Hotspot tier1 on QEMU has no new errors again after the changes.

Thank you all for taking the time to review this!

/integrate

RealFYang commented:
/sponsor
